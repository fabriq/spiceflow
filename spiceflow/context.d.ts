import * as context from './dist/context';
export = context;
