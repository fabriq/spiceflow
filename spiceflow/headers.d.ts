import * as headers from './dist/headers';
export = headers;
