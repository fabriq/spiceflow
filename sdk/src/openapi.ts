import { OpenAPIV3 } from 'openapi-types'
import fs from 'fs'
import fastJsonStableStringify from 'fast-json-stable-stringify'

interface SchemaLocation {
  schema: OpenAPIV3.SchemaObject
  path: string
  method?: string
  type: 'parameter' | 'requestBody' | 'response'
  contentType?: string
  responseCode?: string
  // Add reference to the object that contains the schema
  schemaContainer: {
    schema: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject
  }
}


export function cleanupOpenApi(
  obj: any,
  maxLength: number = 400,
  path: string[] = [],
  removedPaths: string[] = []
): any {
  if (obj === null || typeof obj !== 'object') {
    // Check if string is too long
    if (typeof obj === 'string' && obj.length > maxLength) {
      removedPaths.push(path.join('.'))
      return undefined
    }
    return obj
  }

  // Handle arrays
  if (Array.isArray(obj)) {
    const newArray = obj
      .map((item, index) =>
        cleanupOpenApi(item, maxLength, [...path, index.toString()], removedPaths)
      )
      .filter((x) => x !== undefined)
    return newArray.length > 0 ? newArray : undefined
  }

  // Handle objects
  const newObj: any = {}
  let hasValidProps = false

  for (const [key, value] of Object.entries(obj)) {
    const cleaned = cleanupOpenApi(value, maxLength, [...path, key], removedPaths)
    if (cleaned !== undefined) {
      newObj[key] = cleaned
      hasValidProps = true
    }
  }

  return hasValidProps ? newObj : undefined
}



export function getSubSchemas(
  schemas: SchemaLocation[],
  maxDepth: number = 5,
): SchemaLocation[] {
  const result: SchemaLocation[] = []
  const visited = new Set<string>()

  function processSchema(
    schema: OpenAPIV3.SchemaObject,
    location: SchemaLocation,
    depth: number,
  ) {
    if (depth >= maxDepth) return

    // Generate unique key for schema to avoid cycles
    const schemaKey = fastJsonStableStringify(schema)
    if (visited.has(schemaKey)) return
    visited.add(schemaKey)

    result.push(location)

    if (schema.type === 'object' && schema.properties) {
      for (const [propName, propSchema] of Object.entries(schema.properties)) {
        if ('type' in propSchema) {
          processSchema(
            propSchema,
            {
              ...location,
              path: `${location.path}/properties/${propName}`,
              schema: propSchema,
              schemaContainer: { schema: propSchema },
            },
            depth + 1,
          )
        }
      }
    }

    if (schema.type === 'array' && schema.items) {
      if (!('$ref' in schema.items)) {
        processSchema(
          schema.items,
          {
            ...location,
            path: `${location.path}/items`,
            schema: schema.items,
            schemaContainer: { schema: schema.items },
          },
          depth + 1,
        )
      }
    }

    // Handle allOf, anyOf, oneOf
    ;['allOf', 'anyOf', 'oneOf'].forEach((combiner) => {
      if (schema[combiner as keyof OpenAPIV3.SchemaObject]) {
        const schemas = schema[
          combiner as keyof OpenAPIV3.SchemaObject
        ] as OpenAPIV3.SchemaObject[]
        schemas.forEach((subSchema, index) => {
          if ('type' in subSchema) {
            processSchema(
              subSchema,
              {
                ...location,
                path: `${location.path}/${combiner}/${index}`,
                schema: subSchema,
                schemaContainer: { schema: subSchema },
              },
              depth + 1,
            )
          }
        })
      }
    })

    // Handle additionalProperties
    if (
      schema.additionalProperties &&
      typeof schema.additionalProperties === 'object'
    ) {
      if (!('$ref' in schema.additionalProperties)) {
        processSchema(
          schema.additionalProperties,
          {
            ...location,
            path: `${location.path}/additionalProperties`,
            schema: schema.additionalProperties,
            schemaContainer: { schema: schema.additionalProperties },
          },
          depth + 1,
        )
      }
    }
  }

  schemas.forEach((location) => {
    processSchema(location.schema, location, 0)
  })

  return result
}

function getAllSchemas(openapi: OpenAPIV3.Document): SchemaLocation[] {
  const schemas: SchemaLocation[] = []

  if (!openapi.paths) {
    return schemas
  }

  for (const [path, pathItem] of Object.entries(openapi.paths)) {
    if (!pathItem || Array.isArray(pathItem)) {
      continue
    }

    for (const [method, operation] of Object.entries(pathItem)) {
      if (
        !operation ||
        typeof operation !== 'object' ||
        Array.isArray(operation)
      ) {
        continue
      }

      const op = operation as OpenAPIV3.OperationObject

      // Get request body schemas
      if (op.requestBody && !Array.isArray(op.requestBody)) {
        const requestBody = op.requestBody as OpenAPIV3.RequestBodyObject
        if (requestBody.content) {
          for (const [contentType, mediaType] of Object.entries(
            requestBody.content,
          )) {
            if (mediaType.schema && !('$ref' in mediaType.schema)) {
              schemas.push({
                schema: mediaType.schema,
                path,
                method,
                type: 'requestBody',
                contentType,
                schemaContainer: mediaType as any,
              })
            }
          }
        }
      }

      // Get response schemas
      if (op.responses) {
        for (const [responseCode, response] of Object.entries(op.responses)) {
          if (response && 'content' in response) {
            const responseObj = response as OpenAPIV3.ResponseObject
            if (responseObj.content) {
              for (const [contentType, mediaType] of Object.entries(
                responseObj.content,
              )) {
                if (mediaType.schema && !('$ref' in mediaType.schema)) {
                  schemas.push({
                    schema: mediaType.schema,
                    path,
                    method,
                    type: 'response',
                    contentType,
                    responseCode,
                    schemaContainer: mediaType as any,
                  })
                }
              }
            }
          }
        }
      }

      // Get parameter schemas
      if (op.parameters) {
        for (const param of op.parameters) {
          if (!Array.isArray(param) && 'schema' in param) {
            const paramObj = param as OpenAPIV3.ParameterObject
            if (paramObj.schema && !('$ref' in paramObj.schema)) {
              schemas.push({
                schema: paramObj.schema,
                path,
                method,
                type: 'parameter',
                schemaContainer: paramObj as any,
              })
            }
          }
        }
      }
    }
  }

  return schemas
}

export function deduplicateSchemas(
  openapi: OpenAPIV3.Document,
  getSchemaName: (schema: OpenAPIV3.SchemaObject, index: number) => string = (
    _,
    i,
  ) => `Schema${i + 1}`,
) {
  // Initialize components.schemas if they don't exist
  if (!openapi.components) {
    openapi.components = {}
  }
  if (!openapi.components.schemas) {
    openapi.components.schemas = {}
  }

  const schemasTopLevel = getAllSchemas(openapi)
  const skip = new Set<OpenAPIV3.NonArraySchemaObjectType>([
    'number',
    'string',
    'boolean',
    'integer',
  ])
  const schemas = getSubSchemas(schemasTopLevel).filter((x) => {
    const schema = x.schema

    return schema.enum || !skip.has(schema.type as any)
  })

  const schemaHashes = new Map<
    string,
    {
      name: string
      count: number
      schema: OpenAPIV3.SchemaObject
    }
  >()

  // First pass - hash all schemas and count duplicates
  for (const { schema } of schemas) {
    const hash = fastJsonStableStringify(schema)
    const existing = schemaHashes.get(hash)
    if (existing) {
      existing.count++
    } else {
      schemaHashes.set(hash, {
        name: getSchemaName(schema, schemaHashes.size),
        count: 1,
        schema,
      })
    }
  }

  console.log(
    `Found schemas to deduplicate:`,
    [...schemaHashes.values()].filter((x) => x.count > 1).length,
  )
  // Second pass - replace duplicates with refs
  for (const [hash, { name, count, schema }] of schemaHashes) {
    if (count > 1) {
      // Add schema to components
      openapi.components!.schemas![name] = schema
      const ref = { $ref: `#/components/schemas/${name}` }

      // Replace all matching schemas with ref
      for (const location of schemas) {
        if (fastJsonStableStringify(location.schema) !== hash) {
          continue
        }

        // Simply update the schema reference in the container
        location.schemaContainer.schema = ref
      }
    }
  }

  return { openapi, schemas, schemasTopLevel }
}
