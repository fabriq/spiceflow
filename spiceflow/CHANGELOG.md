# spiceflow

## 0.0.3

### Patch Changes

- Fix \_\_dirname

## 0.0.2

### Patch Changes

- Added experimental --openapi

## 0.0.1

### Patch Changes

- 0eef621: Initial release
