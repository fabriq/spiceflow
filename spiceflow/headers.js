module.exports = require('./dist/headers');
