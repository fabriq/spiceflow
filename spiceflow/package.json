{
  "name": "spiceflow",
  "version": "1.2.0",
  "description": "Simple API framework with RPC and type safety",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "repository": "https://github.com/remorses/spiceflow",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./cors": {
      "types": "./dist/cors.d.ts",
      "default": "./dist/cors.js"
    },
    "./mcp": {
      "types": "./dist/mcp.d.ts",
      "default": "./dist/mcp.js"
    },
    "./client": {
      "types": "./dist/client/index.d.ts",
      "default": "./dist/client/index.js"
    },
    "./openapi": {
      "types": "./dist/openapi.d.ts",
      "default": "./dist/openapi.js"
    },
    "./dist/*": {
      "types": "./dist/*.d.ts",
      "default": "./dist/*.js"
    }
  },
  "scripts": {
    "build": "cp ../README.md ./README.md && rm -rf dist && tsc",
    "prepublishOnly": "pnpm build",
    "gen-openapi": "pnpm tsx scripts/openapi-fern.ts",
    "docs:dev": "pnpm run gen-openapi && fern docs dev",
    "docs:prod": "pnpm run gen-openapi && fern generate --docs --force",
    "fern-sdk": "pnpm run gen-openapi && fern generate --force ",
    "fern-docs": "pnpm run gen-openapi && fern generate --docs --force",
	"play-sdk:build": "pnpm vite build --config ./scripts/play-sdk.vite.ts",
    "test": "pnpm vitest",
    "prepare": "pnpm build",
    "watch": "tsc -w"
  },
  "files": [
    "dist",
    "src"
  ],
  "keywords": [],
  "author": "Tommaso De Rossi, morse <beats.by.morse@gmail.com>",
  "license": "",
  "dependencies": {
    "@medley/router": "^0.2.1",
    "@modelcontextprotocol/sdk": "^1.0.4",
    "@sinclair/typebox": "^0.33.7",
    "ajv": "^8.17.1",
    "ajv-formats": "^3.0.1",
    "eventsource-parser": "^2.0.1",
    "fast-querystring": "^1.1.2",
    "lodash.clonedeep": "^4.5.0",
    "openapi-types": "^12.1.3",
    "zod": "^3.23.8",
    "zod-to-json-schema": "^3.24.1"
  },
  "devDependencies": {
    "@types/node": "22.7.4",
    "eventsource": "^3.0.2",
    "formdata-node": "^6.0.3",
    "js-base64": "^3.7.7",
    "js-yaml": "^4.1.0"
  }
}
