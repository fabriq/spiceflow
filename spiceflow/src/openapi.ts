/* eslint-disable @typescript-eslint/ban-ts-comment */
import { JSONSchemaType } from 'ajv'
import { InternalRoute, Spiceflow } from '.'

import type { OpenAPIV3 } from 'openapi-types'

let excludeMethods = ['OPTIONS']
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-unused-vars */
import type { HTTPMethod, LocalHook, TypeSchema } from './elysia-fork/types'

import { Kind, type TSchema } from '@sinclair/typebox'

import deepClone from 'lodash.clonedeep'
import { z } from 'zod'
import zodToJsonSchema from 'zod-to-json-schema'

export const toOpenAPIPath = (path: string) =>
	path
		.split('/')
		.map((x) => {
			if (x.startsWith(':')) {
				x = x.slice(1, x.length)
				if (x.endsWith('?')) x = x.slice(0, -1)
				x = `{${x}}`
			}

			return x
		})
		.join('/')

export const mapProperties = (
	name: string,
	schema: TypeSchema | string | undefined,
	models: Record<string, TypeSchema>,
) => {
	if (schema === undefined) return []

	if (typeof schema === 'string')
		if (schema in models) schema = models[schema]
		else throw new Error(`Can't find model ${schema}`)

	let jsonSchema = getJsonSchema(schema)

	return Object.entries(jsonSchema?.properties ?? []).map(([key, value]) => {
		const {
			type: valueType = undefined,
			description,
			examples,
			...schemaKeywords
		} = value as any
		return {
			// @ts-ignore
			description,
			examples,
			schema: { type: valueType, ...schemaKeywords },
			in: name,
			name: key,

			required: jsonSchema!.required?.includes(key) ?? false,
		}
	})
}

const mapTypesResponse = (
	types: string[],
	schema:
		| string
		| {
				type: string
				properties: Object
				required: string[]
		  },
) => {
	if (
		typeof schema === 'object' &&
		['void', 'undefined', 'null'].includes(schema.type)
	)
		return

	const responses: Record<string, OpenAPIV3.MediaTypeObject> = {}

	for (const type of types)
		responses[type] = {
			schema:
				typeof schema === 'string'
					? {
							$ref: `#/components/schemas/${schema}`,
					  }
					: { ...(schema as any) },
		}

	return responses
}

export const capitalize = (word: string) =>
	word.charAt(0).toUpperCase() + word.slice(1)

export const generateOperationId = (method: string, paths: string) => {
	let operationId = method.toLowerCase()

	if (paths === '/') return operationId + 'Index'

	for (const path of paths.split('/')) {
		if (path.charCodeAt(0) === 123) {
			operationId += 'By' + capitalize(path.slice(1, -1))
		} else {
			operationId += capitalize(path)
		}
	}

	return operationId
}

export const registerSchemaPath = ({
	schema,
	path,
	method,
	hook,
	models,
}: {
	schema: Partial<OpenAPIV3.PathsObject>
	contentType?: string | string[]
	path: string
	method: HTTPMethod
	hook?: LocalHook<any, any, any, any, any, any, any>
	models: Record<string, TypeSchema>
}) => {
	if (hook) hook = deepClone(hook)

	// TODO if a route uses an async generator, add text/event-stream. if a roue does not add an explicit schema, use all possible content types
	const contentType = hook?.type ?? [
		'application/json',
		// 'multipart/form-data',
		// 'text/plain',
	]

	path = toOpenAPIPath(path)

	const contentTypes =
		typeof contentType === 'string'
			? [contentType]
			: contentType ?? ['application/json']

	const bodySchema = hook?.body
	const paramsSchema = hook?.params
	// const headerSchema = hook?.headers
	const querySchema = hook?.query
	let responseSchema = hook?.response as unknown as OpenAPIV3.ResponsesObject

	if (typeof responseSchema === 'object') {
		if (Kind in responseSchema) {
			const {
				type,
				properties,
				required,
				additionalProperties,
				patternProperties,
				...rest
			} = responseSchema as typeof responseSchema & {
				type: string
				properties: Object
				required: string[]
			}

			responseSchema = {
				'200': {
					...rest,
					description: rest.description as any,
					content: mapTypesResponse(
						contentTypes,
						type === 'object' || type === 'array'
							? ({
									type,
									properties,
									patternProperties,
									items: responseSchema.items,
									required,
							  } as any)
							: responseSchema,
					),
				},
			}
		} else {
			Object.entries(responseSchema as Record<string, TSchema>).forEach(
				([key, value]) => {
					if (typeof value === 'string') {
						if (!models[value]) return

						// eslint-disable-next-line @typescript-eslint/no-unused-vars
						const {
							type,
							properties,
							required,
							additionalProperties: _1,
							patternProperties: _2,
							...rest
						} = models[value] as TSchema & {
							type: string
							properties: Object
							required: string[]
						}

						responseSchema[key] = {
							...rest,
							description: rest.description as any,
							content: mapTypesResponse(contentTypes, value),
						}
					} else {
						const {
							type,
							properties,
							required,
							additionalProperties,
							patternProperties,
							...rest
						} = value as typeof value & {
							type: string
							properties: Object
							required: string[]
						}

						responseSchema[key] = {
							...rest,
							description: rest.description as any,
							content: mapTypesResponse(
								contentTypes,
								type === 'object' || type === 'array'
									? ({
											type,
											properties,
											patternProperties,
											items: value.items,
											required,
									  } as any)
									: value,
							),
						}
					}
				},
			)
		}
	} else if (typeof responseSchema === 'string') {
		if (!(responseSchema in models)) return

		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const {
			type,
			properties,
			required,
			additionalProperties: _1,
			patternProperties: _2,
			...rest
		} = models[responseSchema] as TSchema & {
			type: string
			properties: Object
			required: string[]
		}

		responseSchema = {
			// @ts-ignore
			'200': {
				...rest,
				content: mapTypesResponse(contentTypes, responseSchema),
			},
		}
	}

	const parameters = [
		// ...mapProperties('header', headerSchema, models),
		...mapProperties('path', paramsSchema, models),
		...mapProperties('query', querySchema, models),
	]

	schema[path] = {
		...(schema[path] ? schema[path] : {}),
		[method.toLowerCase()]: {
			...((paramsSchema || querySchema || bodySchema
				? ({ parameters } as any)
				: {}) satisfies OpenAPIV3.ParameterObject),
			...(responseSchema
				? {
						responses: responseSchema,
				  }
				: {}),
			operationId:
				hook?.detail?.operationId ?? generateOperationId(method, path),
			...hook?.detail,
			...(bodySchema
				? {
						requestBody: {
							required: true,
							content: mapTypesResponse(
								contentTypes,
								typeof bodySchema === 'string'
									? {
											$ref: `#/components/schemas/${bodySchema}`,
									  }
									: (bodySchema as any),
							),
						},
				  }
				: null),
		} satisfies OpenAPIV3.OperationObject,
	}
}

/**
 * Plugin for [elysia](https://github.com/elysiajs/elysia) that auto-generate Swagger page.
 *
 * @see https://github.com/elysiajs/elysia-swagger
 */
export const openapi = <Path extends string = '/openapi'>({
	path,
	documentation = {},
}: {
	path: Path
	/**
	 * Customize Swagger config, refers to Swagger 2.0 config
	 *
	 * @see https://swagger.io/specification/v2/
	 */
	documentation?: Omit<
		Partial<OpenAPIV3.Document>,
		| 'x-express-openapi-additional-middleware'
		| 'x-express-openapi-validation-strict'
	>
}) => {
	const schema = {}
	let totalRoutes = 0

	const relativePath = path.startsWith('/') ? path.slice(1) : path

	const app = new Spiceflow({ name: 'openapi' }).get(path, ({}) => {
		let routes = app.getAllRoutes()
		if (routes.length !== totalRoutes) {
			const ALLOWED_METHODS = [
				'GET',
				'PUT',
				'POST',
				'DELETE',
				'OPTIONS',
				'HEAD',
				'PATCH',
				'TRACE',
			]
			totalRoutes = routes.length

			routes.forEach((route: InternalRoute) => {
				if (route.hooks?.detail?.hide === true) return
				// TODO: route.hooks?.detail?.hide !== false  add ability to hide: false to prevent excluding
				if (excludeMethods.includes(route.method)) return
				if (
					ALLOWED_METHODS.includes(route.method) === false &&
					route.method !== 'ALL'
				)
					return

				if (route.method === 'ALL') {
					ALLOWED_METHODS.forEach((method) => {
						registerSchemaPath({
							schema,
							hook: route.hooks,
							method,
							path: route.path,
							// @ts-ignore
							models: app.definitions?.type,
							contentType: route.hooks.type,
						})
					})
					return
				}

				registerSchemaPath({
					schema,
					hook: route.hooks,
					method: route.method,
					path: route.path,
					// @ts-ignore
					models: app.definitions?.type,
					contentType: route.hooks.type,
				})
			})
		}

		return {
			openapi: '3.0.3',
			...{
				...documentation,
				// tags: documentation.tags?.filter(
				// 	(tag) => !excludeTags?.includes(tag?.name),
				// ),
				info: {
					title: 'Elysia Documentation',
					description: 'Development documentation',
					version: '0.0.0',
					...documentation.info,
				},
			},
			paths: {
				...schema,
				...documentation.paths,
			},
			components: {
				...documentation.components,
				schemas: {
					// @ts-ignore
					...app.definitions?.type,
					...documentation.components?.schemas,
				},
			},
		} satisfies OpenAPIV3.Document
	})

	return app
}

function getJsonSchema(schema: TypeSchema): JSONSchemaType<any> {
	if (schema instanceof z.ZodType) {
		let jsonSchema = zodToJsonSchema(schema, {})
		return jsonSchema as any
	}

	return schema as any
}
