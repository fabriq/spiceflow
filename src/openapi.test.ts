import { assertSnapshot } from "@std/testing/snapshot";
import { Spiceflow } from "./spiceflow.ts";
import { openapi } from "./openapi.ts";
import { z } from "zod";

Deno.test("openapi response", async (t) => {
  const app = new Spiceflow()
    .use(
      openapi({
        info: {
          title: "Spiceflow Docs",
          version: "0.0.0",
        },
      }),
    )
    .use(
      new Spiceflow({ basePath: "/one" }).get(
        "/ids/:id",
        ({ params }) => {
          if (Math.random() < 0.5) {
            // TODO add a way to set status

            return {
              message: "sdf",
            };
          }
          return params.id;
        },
        {
          response: {
            200: z.string(),
            404: z.object({
              message: z.string(),
            }),
          },
        },
      ),
    )
    .patch(
      "/addBody",
      async (c) => {
        const body = await c.request.json();
        return body;
      },
      {
        body: z.object({
          name: z.string(),
        }),
        response: z.object({
          name: z.string().optional(),
        }),
      },
    )
    .get(
      "/queryParams",
      async (c) => {
        const query = c.query;
        return query;
      },
      {
        query: z.object({
          name: z.string(),
        }),
        response: z.object({
          name: z.string().optional(),
        }),
      },
    )
    .post(
      "/queryParams",
      async (c) => {
        const query = c.query;
        return query;
      },
      {
        detail: {
          description: "This is a post",
          operationId: "postQueryParamsXXX",
        },
        body: z.object({
          name: z.string(),
        }),
        response: z.object({
          name: z.string().optional(),
        }),
      },
    )
    .get(
      "/stream",
      async function* () {
        for (let i = 0; i < 3; i++) {
          yield { count: i };
          await new Promise((resolve) => setTimeout(resolve, 10));
        }
      },
      {
        detail: {
          description: "This is a stream",
        },
        // response: z.object({
        //   count: z.number(),
        // }),
      },
    )
    .get(
      "/streamWithSchema",
      async function* () {
        for (let i = 0; i < 3; i++) {
          yield { count: i };
          await new Promise((resolve) => setTimeout(resolve, 10));
        }
      },
      {
        detail: {
          description: "This is a stream with schema",
        },
        response: z.object({
          count: z.number(),
        }),
      },
    )
    .post(
      "/formWithSchemaForm",
      () => {
        const formData = new FormData();
        formData.append("name", "test");
        formData.append("age", "25");
        return new Response(formData, {
          headers: {
            "content-type": "multipart/form-data",
          },
        });
      },
      {
        detail: {
          description: "This returns form data with schema",
        },
        response: z.object({
          name: z.string(),
          age: z.string(),
        }),
        type: "multipart/form-data",
      },
    )
    .use(
      new Spiceflow({ basePath: "/two" }).get(
        "/ids/:id",
        ({ params }) => params.id,
        {
          params: z.object({
            id: z.string(),
          }),
        },
      ),
    );
  const openapiSchema = await app
    .handle(new Request("http://localhost/openapi"))
    .then((x) => x.json());

  await assertSnapshot(t, openapiSchema);
});
