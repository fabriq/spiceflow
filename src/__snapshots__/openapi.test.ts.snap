export const snapshot = {};

snapshot[`openapi response 1`] = `
{
  __superjsonMeta: {
    values: {
      "paths./addBody.patch.responses.200.content.application/json.schema.items": [
        "undefined",
      ],
      "paths./addBody.patch.responses.200.content.application/json.schema.patternProperties": [
        "undefined",
      ],
      "paths./addBody.patch.responses.200.content.application/json.schema.required": [
        "undefined",
      ],
      "paths./formWithSchemaForm.post.responses.200.content.multipart/form-data.schema.items": [
        "undefined",
      ],
      "paths./formWithSchemaForm.post.responses.200.content.multipart/form-data.schema.patternProperties": [
        "undefined",
      ],
      "paths./one/ids/{id}.get.parameters.0.description": [
        "undefined",
      ],
      "paths./one/ids/{id}.get.parameters.0.examples": [
        "undefined",
      ],
      "paths./one/ids/{id}.get.responses.404.content.application/json.schema.items": [
        "undefined",
      ],
      "paths./one/ids/{id}.get.responses.404.content.application/json.schema.patternProperties": [
        "undefined",
      ],
      "paths./queryParams.get.parameters.0.description": [
        "undefined",
      ],
      "paths./queryParams.get.parameters.0.examples": [
        "undefined",
      ],
      "paths./queryParams.get.responses.200.content.application/json.schema.items": [
        "undefined",
      ],
      "paths./queryParams.get.responses.200.content.application/json.schema.patternProperties": [
        "undefined",
      ],
      "paths./queryParams.get.responses.200.content.application/json.schema.required": [
        "undefined",
      ],
      "paths./queryParams.post.responses.200.content.application/json.schema.items": [
        "undefined",
      ],
      "paths./queryParams.post.responses.200.content.application/json.schema.patternProperties": [
        "undefined",
      ],
      "paths./queryParams.post.responses.200.content.application/json.schema.required": [
        "undefined",
      ],
      "paths./streamWithSchema.get.responses.200.content.application/json.schema.items": [
        "undefined",
      ],
      "paths./streamWithSchema.get.responses.200.content.application/json.schema.patternProperties": [
        "undefined",
      ],
      "paths./two/ids/{id}.get.parameters.0.description": [
        "undefined",
      ],
      "paths./two/ids/{id}.get.parameters.0.examples": [
        "undefined",
      ],
    },
  },
  components: {
    schemas: {},
  },
  info: {
    description: "Development documentation",
    title: "Spiceflow Docs",
    version: "0.0.0",
  },
  openapi: "3.1.3",
  paths: {
    "/addBody": {
      patch: {
        parameters: [],
        requestBody: {
          content: {
            "application/json": {
              schema: {
                additionalProperties: true,
                properties: {
                  name: {
                    type: "string",
                  },
                },
                required: [
                  "name",
                ],
                type: "object",
              },
            },
          },
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    name: {
                      type: "string",
                    },
                  },
                  required: null,
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/formWithSchemaForm": {
      post: {
        description: "This returns form data with schema",
        responses: {
          "200": {
            content: {
              "multipart/form-data": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    age: {
                      type: "string",
                    },
                    name: {
                      type: "string",
                    },
                  },
                  required: [
                    "name",
                    "age",
                  ],
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/one/ids/{id}": {
      get: {
        parameters: [
          {
            description: null,
            examples: null,
            in: "path",
            name: "id",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  type: "string",
                },
              },
            },
            description: "",
          },
          "404": {
            content: {
              "application/json": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    message: {
                      type: "string",
                    },
                  },
                  required: [
                    "message",
                  ],
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/openapi": {
      get: {
        responses: {
          "200": {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/queryParams": {
      get: {
        parameters: [
          {
            description: null,
            examples: null,
            in: "query",
            name: "name",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    name: {
                      type: "string",
                    },
                  },
                  required: null,
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
      post: {
        description: "This is a post",
        operationId: "postQueryParamsXXX",
        parameters: [],
        requestBody: {
          content: {
            "application/json": {
              schema: {
                additionalProperties: true,
                properties: {
                  name: {
                    type: "string",
                  },
                },
                required: [
                  "name",
                ],
                type: "object",
              },
            },
          },
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    name: {
                      type: "string",
                    },
                  },
                  required: null,
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/stream": {
      get: {
        "x-fern-streaming": {
          format: "sse",
        },
        description: "This is a stream",
        responses: {
          "200": {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/streamWithSchema": {
      get: {
        "x-fern-streaming": {
          format: "sse",
        },
        description: "This is a stream with schema",
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: null,
                  patternProperties: null,
                  properties: {
                    count: {
                      type: "number",
                    },
                  },
                  required: [
                    "count",
                  ],
                  type: "object",
                },
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
    "/two/ids/{id}": {
      get: {
        parameters: [
          {
            description: null,
            examples: null,
            in: "path",
            name: "id",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {
          "200": {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
          default: {
            content: {
              "*/*": {
                schema: {},
              },
            },
            description: "",
          },
        },
      },
    },
  },
}
`;
